pkg/antctl/client.go

```
@@ -0,0 +1,126 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+)
+
+var isAgent = strings.HasPrefix(os.Getenv("POD_NAME"), "antrea-agent")
+
+// RequestOption contains options to do a request to CLI server.
+type RequestOption struct {
+	Kubeconfig string
+	Name       string
+	Args       map[string]string
+}
+
+// URI returns the request path of the request option.
+func (opt *RequestOption) URI(gv *schema.GroupVersion) string {
+	u := new(url.URL)
+	u.Path = path.Join("/apis", gv.Group, gv.Version, opt.Name)
+
+	q := u.Query()
+	for k, v := range opt.Args {
+		q.Set(k, v)
+	}
+	u.RawQuery = q.Encode()
+
+	return u.RequestURI()
+}
+
+// Client does request to a CLI server and get the response.
+type Client struct {
+	GroupVersion *schema.GroupVersion
+	Codec        serializer.CodecFactory
+}
+
+func (c *Client) resolveKubeconfig(opt *RequestOption) (*rest.Config, error) {
+	kubeconfig, err := clientcmd.BuildConfigFromFlags("", opt.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeconfig.GroupVersion = c.GroupVersion
+	kubeconfig.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: c.Codec}
+	return kubeconfig, nil
+}
+
+func (c *Client) localDo(cmd *cobra.Command, opt *RequestOption) (io.Reader, error) {
+	client := &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, _, _ string) (conn net.Conn, err error) {
+				return new(net.Dialer).DialContext(ctx, "unix", unixDomainSockAddr)
+			},
+		},
+	}
+	uri := opt.URI(c.GroupVersion)
+	klog.Infof("Requesting URI %s", uri)
+	resp, err := client.Get("http://antctl" + uri)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	var buf bytes.Buffer
+	io.Copy(&buf, resp.Body)
+	return &buf, nil
+}
+
+// Do makes the request by the command and the request option.
+func (c *Client) Do(cmd *cobra.Command, opt *RequestOption) (io.Reader, error) {
+	kubeconfig, err := c.resolveKubeconfig(opt)
+	if err != nil {
```

nit: Is this the right test here? Is it possible that I have a malformed kubeconfig file, in which case we should probably return an error instead of falling back to local mode.

---

Make sense. I changed to you said. See `pkg/antctl/client.go:126` and `pkg/antctl/client.go:94`.